File: apps/android/app/src/main/java/dev/flextranslate/foundation/ModelDownloadEngine.kt
package dev.flextranslate.foundation

import java.io.File
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Чистый, без Android, движок загрузки одного пака модели. Держит всю реальную работу с сетью и
 * диском, чтобы её можно было детерминированно гонять в юнит-тестах против локального HTTP-сервера:
 *
 *  - **Докачка**: недокачанный `<file>.part` продолжаем запросом `Range: bytes=<have>-`; если сервер
 *    его уважает (206) — дописываем, иначе (200) перекачиваем файл с нуля.
 *  - **Атомарная установка**: байты пишутся в `<file>.part`, проверяются по SHA-256 и только при
 *    совпадении `renameTo` в финальное имя — читатель никогда не видит недописанный или битый файл.
 *  - **Откат**: при несовпадении размера/чек-суммы или ошибке I/O `.part` удаляется, чтобы ретрай
 *    стартовал с чистого листа (докачка Range с битого хвоста никогда не сошлась бы).
 *  - **Идемпотентность**: уже лежащий файл с верным размером И чек-суммой пропускаем.
 *  - **Отменяемость**: общий [AtomicBoolean] опрашивается между чанками; отмена оставляет `.part`
 *    на месте, чтобы следующая попытка докачала, а не качала заново.
 *
 * Движок не лезет в connectivity и UI — это надстраивает поверх [ModelDownloadManager].
 */
class ModelDownloadEngine(
    private val connectTimeoutMs: Int = DEFAULT_CONNECT_TIMEOUT_MS,
    private val readTimeoutMs: Int = DEFAULT_READ_TIMEOUT_MS,
) {

    /** Суммарный прогресс по файлам пака. [bytesDone] включает уже установленные файлы. */
    data class Progress(
        val bytesDone: Long,
        val bytesTotal: Long,
        val currentFile: String?,
    ) {
        val fraction: Float
            get() = if (bytesTotal <= 0L) 0f else (bytesDone.toFloat() / bytesTotal.toFloat()).coerceIn(0f, 1f)
    }

    /** Итоговый исход загрузки пака. */
    sealed interface Result {
        data object Success : Result
        data object Cancelled : Result
        data class Failure(val message: String) : Result
    }

    /**
     * Качает каждый файл из [spec] в [modelDir], дёргая [onProgress] по мере накопления байт.
     * [cancelled] опрашивается кооперативно. Возвращает терминальный [Result]; на ожидаемых
     * сбоях сети/IO/проверки не бросает исключений — они становятся [Result.Failure].
     */
    fun download(
        spec: ModelDownloadSpec,
        modelDir: File,
        cancelled: AtomicBoolean,
        onProgress: (Progress) -> Unit,
    ): Result {
        if (!modelDir.exists() && !modelDir.mkdirs()) {
            return Result.Failure("Не удалось создать каталог модели: ${modelDir.path}")
        }
        val total = spec.totalBytes
        // Файлы, уже подтверждённые на входе, сразу идут в зачёт прогресса (идемпотентная докачка).
        var completedBytes = spec.files.sumOf { file ->
            if (isInstalledAndVerified(File(modelDir, file.fileName), file)) file.sizeBytes else 0L
        }
        onProgress(Progress(completedBytes, total, null))

        for (file in spec.files) {
            if (cancelled.get()) return Result.Cancelled
            val target = File(modelDir, file.fileName)
            if (isInstalledAndVerified(target, file)) continue

            val baseBytes = completedBytes
            val result = downloadFile(file, target, cancelled) { fileBytes ->
                onProgress(Progress(baseBytes + fileBytes, total, file.fileName))
            }
            when (result) {
                is FileResult.Done -> completedBytes = baseBytes + file.sizeBytes
                FileResult.Cancelled -> return Result.Cancelled
                is FileResult.Failed -> return Result.Failure(result.message)
            }
            onProgress(Progress(completedBytes, total, null))
        }
        return Result.Success
    }

    private sealed interface FileResult {
        data object Done : FileResult
        data object Cancelled : FileResult
        data class Failed(val message: String) : FileResult
    }

    private fun downloadFile(
        file: ModelFileDownload,
        target: File,
        cancelled: AtomicBoolean,
        onFileBytes: (Long) -> Unit,
    ): FileResult {
        val part = File(target.parentFile, target.name + PART_SUFFIX)
        // Залежавшийся .part больше ожидаемого размера — битый хвост; сносим перед докачкой.
        if (part.exists() && part.length() > file.sizeBytes) part.delete()
        val existing = if (part.isFile) part.length() else 0L

        val connection = try {
            openConnection(file.sourceUrl, existing)
        } catch (io: IOException) {
            return FileResult.Failed("Сеть недоступна: ${io.message ?: file.fileName}")
        }

        return try {
            val status = connection.responseCode
            val resuming = existing > 0L && status == HttpURLConnection.HTTP_PARTIAL
            // Сервер проигнорил Range → качаем файл с нуля, не дописываем.
            val startFrom = if (resuming) existing else 0L
            if (startFrom == 0L && part.exists()) part.delete()
            if (status != HttpURLConnection.HTTP_OK && status != HttpURLConnection.HTTP_PARTIAL) {
                return FileResult.Failed("HTTP $status для ${file.fileName}")
            }
            streamToPart(connection, part, file, startFrom, cancelled, onFileBytes)
        } finally {
            connection.disconnect()
        }
    }

    private fun streamToPart(
        connection: HttpURLConnection,
        part: File,
        file: ModelFileDownload,
        startFrom: Long,
        cancelled: AtomicBoolean,
        onFileBytes: (Long) -> Unit,
    ): FileResult {
        var written = startFrom
        try {
            connection.inputStream.use { input ->
                appendTo(part, startFrom > 0L).use { output ->
                    val buffer = ByteArray(BUFFER_BYTES)
                    while (true) {
                        if (cancelled.get()) return FileResult.Cancelled
                        val read = input.read(buffer)
                        if (read < 0) break
                        output.write(buffer, 0, read)
                        written += read
                        onFileBytes(written)
                    }
                }
            }
        } catch (io: IOException) {
            // Сеть оборвалась посреди потока: оставляем .part, чтобы потом докачать отсюда.
            return FileResult.Failed("Ошибка загрузки ${file.fileName}: ${io.message ?: "I/O"}")
        }

        if (written != file.sizeBytes) {
            part.delete() // размер не сошёлся → битый, откатываем, чтобы ретрай был с чистого листа.
            return FileResult.Failed("Размер ${file.fileName} не совпал ($written≠${file.sizeBytes})")
        }
        val actual = sha256Of(part)
        if (!actual.equals(file.sha256, ignoreCase = true)) {
            part.delete() // чек-сумма не сошлась → битый, откатываем.
            return FileResult.Failed("Контрольная сумма ${file.fileName} не совпала")
        }
        val target = File(part.parentFile, part.name.removeSuffix(PART_SUFFIX))
        target.delete()
        if (!part.renameTo(target)) {
            part.delete()
            return FileResult.Failed("Не удалось переименовать ${file.fileName}")
        }
        return FileResult.Done
    }

    /** Файл считается установленным только если он есть, верного размера и прошёл чек-сумму. */
    private fun isInstalledAndVerified(target: File, file: ModelFileDownload): Boolean =
        target.isFile && target.length() == file.sizeBytes &&
            sha256Of(target).equals(file.sha256, ignoreCase = true)

    private fun openConnection(url: String, resumeFrom: Long): HttpURLConnection {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.connectTimeout = connectTimeoutMs
        connection.readTimeout = readTimeoutMs
        connection.instanceFollowRedirects = true
        connection.requestMethod = "GET"
        if (resumeFrom > 0L) connection.setRequestProperty("Range", "bytes=$resumeFrom-")
        connection.connect()
        return connection
    }

    private fun appendTo(file: File, append: Boolean) = java.io.FileOutputStream(file, append)

    private fun sha256Of(file: File): String {
        val digest = MessageDigest.getInstance("SHA-256")
        file.inputStream().use { input -> digestStream(input, digest) }
        return digest.digest().joinToString("") { byte -> "%02x".format(byte) }
    }

    private fun digestStream(input: InputStream, digest: MessageDigest) {
        val buffer = ByteArray(BUFFER_BYTES)
        while (true) {
            val read = input.read(buffer)
            if (read <= 0) break
            digest.update(buffer, 0, read)
        }
    }

    companion object {
        const val PART_SUFFIX = ".part"
        private const val BUFFER_BYTES = 1 shl 16
        private const val DEFAULT_CONNECT_TIMEOUT_MS = 15_000
        private const val DEFAULT_READ_TIMEOUT_MS = 30_000
    }
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/ModelDownloadEngine.swift
import Foundation
import CryptoKit

/// Pure, UI-free download engine for one model pack. Owns the real network and disk mechanics
/// so they can be unit-tested deterministically against a local HTTP server:
///
/// - **Resume**: a partial `<file>.part` is continued with an HTTP `Range: bytes=<have>-` request;
///   a server honouring it (206) appends, otherwise (200) the engine restarts that file cleanly.
/// - **Atomic install**: bytes land in `<file>.part`, are SHA-256 verified, then moved to the
///   final name only on a verified match — a reader never sees a half-written or corrupt file.
/// - **Rollback**: on size/checksum mismatch the `.part` is deleted so a retry starts clean.
/// - **Idempotent**: a file already present, correctly sized AND checksum-verified is skipped.
/// - **Cancellable**: cancelled() is polled before each file; cancel leaves the `.part` in place.
///
/// Mirrors Android ModelDownloadEngine.
final class ModelDownloadEngine: @unchecked Sendable {

    static let partSuffix = ".part"
    private let connectTimeoutSeconds: Double
    private let readTimeoutSeconds: Double

    init(connectTimeoutSeconds: Double = 15, readTimeoutSeconds: Double = 30) {
        self.connectTimeoutSeconds = connectTimeoutSeconds
        self.readTimeoutSeconds = readTimeoutSeconds
    }

    /// Aggregate progress across the pack's files.
    struct Progress: Sendable {
        let bytesDone: Int64
        let bytesTotal: Int64
        let currentFile: String?

        var fraction: Float {
            guard bytesTotal > 0 else { return 0 }
            return Float(bytesDone) / Float(bytesTotal)
        }
    }

    /// Terminal outcome of a pack download.
    enum Result: Sendable, Equatable {
        case success
        case cancelled
        case failure(String)
    }

    private enum FileResult {
        case done
        case cancelled
        case failed(String)
    }

    /// Download every file in spec into modelDir.
    /// cancelled() is checked cooperatively before each file.
    /// Never throws; network/IO/verify failures become .failure.
    func download(
        spec: ModelDownloadSpec,
        modelDir: URL,
        cancelled: @escaping () -> Bool,
        onProgress: @escaping (Progress) -> Void
    ) -> Result {
        let fm = FileManager.default
        if !fm.fileExists(atPath: modelDir.path) {
            do {
                try fm.createDirectory(at: modelDir, withIntermediateDirectories: true)
            } catch {
                return .failure("Could not create model dir: \(error.localizedDescription)")
            }
        }
        let total = spec.totalBytes
        var completedBytes: Int64 = spec.files.reduce(0) { acc, file in
            let target = modelDir.appendingPathComponent(file.fileName)
            return acc + (isInstalledAndVerified(target: target, file: file) ? file.sizeBytes : 0)
        }
        onProgress(Progress(bytesDone: completedBytes, bytesTotal: total, currentFile: nil))

        for file in spec.files {
            if cancelled() { return .cancelled }
            let target = modelDir.appendingPathComponent(file.fileName)
            if isInstalledAndVerified(target: target, file: file) { continue }

            let baseBytes = completedBytes
            let fileResult = downloadFile(file: file, target: target, cancelled: cancelled) { fileBytes in
                onProgress(Progress(bytesDone: baseBytes + fileBytes, bytesTotal: total, currentFile: file.fileName))
            }
            switch fileResult {
            case .done:
                completedBytes = baseBytes + file.sizeBytes
            case .cancelled:
                return .cancelled
            case .failed(let message):
                return .failure(message)
            }
            onProgress(Progress(bytesDone: completedBytes, bytesTotal: total, currentFile: nil))
        }
        return .success
    }

    private func downloadFile(
        file: ModelFileDownload,
        target: URL,
        cancelled: @escaping () -> Bool,
        onFileBytes: @escaping (Int64) -> Void
    ) -> FileResult {
        let fm = FileManager.default
        let partURL = URL(fileURLWithPath: target.path + Self.partSuffix)

        // A stale .part larger than the expected size is corrupt — drop it.
        if let attrs = try? fm.attributesOfItem(atPath: partURL.path),
           let size = attrs[.size] as? Int64, size > file.sizeBytes {
            try? fm.removeItem(at: partURL)
        }

        let existing: Int64
        if fm.fileExists(atPath: partURL.path),
           let attrs = try? fm.attributesOfItem(atPath: partURL.path),
           let size = attrs[.size] as? Int64 {
            existing = size
        } else {
            existing = 0
        }

        guard let url = URL(string: file.sourceUrl) else {
            return .failed("Invalid URL: \(file.sourceUrl)")
        }

        var request = URLRequest(url: url, timeoutInterval: readTimeoutSeconds)
        request.httpMethod = "GET"
        if existing > 0 {
            request.setValue("bytes=\(existing)-", forHTTPHeaderField: "Range")
        }

        // Synchronous download via semaphore — engine runs on a background thread.
        let semaphore = DispatchSemaphore(value: 0)
        var responseData: Data?
        var httpStatus: Int = 0
        var networkErrorMsg: String?

        let sessionConfig = URLSessionConfiguration.default
        sessionConfig.timeoutIntervalForRequest = readTimeoutSeconds
        sessionConfig.timeoutIntervalForResource = 3_600
        let session = URLSession(configuration: sessionConfig)

        session.dataTask(with: request) { data, response, error in
            if let error = error {
                networkErrorMsg = error.localizedDescription
            } else if let http = response as? HTTPURLResponse {
                httpStatus = http.statusCode
                responseData = data
            }
            semaphore.signal()
        }.resume()
        semaphore.wait()

        if let msg = networkErrorMsg {
            return .failed("Network error for \(file.fileName): \(msg)")
        }

        let resuming = existing > 0 && httpStatus == 206
        let startFrom: Int64 = resuming ? existing : 0

        if !resuming && existing > 0 {
            // Server ignored Range — restart clean.
            try? fm.removeItem(at: partURL)
        }

        if httpStatus != 200 && httpStatus != 206 {
            return .failed("HTTP \(httpStatus) for \(file.fileName)")
        }

        guard let data = responseData else {
            return .failed("No data received for \(file.fileName)")
        }

        if cancelled() { return .cancelled }

        return streamToPart(
            data: data,
            partURL: partURL,
            target: target,
            file: file,
            startFrom: startFrom,
            onFileBytes: onFileBytes
        )
    }

    private func streamToPart(
        data: Data,
        partURL: URL,
        target: URL,
        file: ModelFileDownload,
        startFrom: Int64,
        onFileBytes: @escaping (Int64) -> Void
    ) -> FileResult {
        let fm = FileManager.default
        let append = startFrom > 0 && fm.fileExists(atPath: partURL.path)

        do {
            if append {
                let handle = try FileHandle(forWritingTo: partURL)
                handle.seekToEndOfFile()
                handle.write(data)
                try handle.close()
            } else {
                try data.write(to: partURL, options: .atomic)
            }
        } catch {
            try? fm.removeItem(at: partURL)
            return .failed("Write error for \(file.fileName): \(error.localizedDescription)")
        }

        guard let attrs = try? fm.attributesOfItem(atPath: partURL.path),
              let written = attrs[.size] as? Int64 else {
            try? fm.removeItem(at: partURL)
            return .failed("Could not stat part file for \(file.fileName)")
        }

        onFileBytes(written)

        if written != file.sizeBytes {
            try? fm.removeItem(at: partURL)
            return .failed("Size mismatch for \(file.fileName) (\(written) != \(file.sizeBytes))")
        }

        // SHA-256 verify.
        guard let partData = try? Data(contentsOf: partURL) else {
            try? fm.removeItem(at: partURL)
            return .failed("Could not read part for SHA-256 verify: \(file.fileName)")
        }
        let digest = SHA256.hash(data: partData)
        let actual = digest.compactMap { String(format: "%02x", $0) }.joined()
        guard actual.lowercased() == file.sha256.lowercased() else {
            try? fm.removeItem(at: partURL)
            return .failed("Checksum mismatch for \(file.fileName)")
        }

        // Atomic rename.
        try? fm.removeItem(at: target)
        do {
            try fm.moveItem(at: partURL, to: target)
        } catch {
            try? fm.removeItem(at: partURL)
            return .failed("Could not rename \(file.fileName): \(error.localizedDescription)")
        }
        return .done
    }

    /// A file counts as installed only when present, correctly sized, and checksum-verified.
    private func isInstalledAndVerified(target: URL, file: ModelFileDownload) -> Bool {
        guard let attrs = try? FileManager.default.attributesOfItem(atPath: target.path),
              let size = attrs[.size] as? Int64, size == file.sizeBytes else { return false }
        guard let data = try? Data(contentsOf: target) else { return false }
        let digest = SHA256.hash(data: data)
        let actual = digest.compactMap { String(format: "%02x", $0) }.joined()
        return actual.lowercased() == file.sha256.lowercased()
    }
}
